### DISCUSSION TITLE
How to filter the intersected points from splitting two meshes

### DISCUSSION THREAD

**Discussion by citystrawman** [CGAL/cgal#8305] [Q&A]
Hi, I used `split `method on two meshes (slope.off and slide.off) interchangeably and got the splitted meshes as follows([Here's ](https://github.com/CGAL/cgal/discussions/8299#discussioncomment-9842107)my previous discussion):

![image](https://github.com/CGAL/cgal/assets/9004748/c38fb66f-adad-494b-9141-2d3b62daf7eb)
![image](https://github.com/CGAL/cgal/assets/9004748/c5de3961-e565-434b-b373-21259e123339)

the green nodes are the new intersected nodes. Then by using `stitch `method I stiched the two meshes into one:
![image](https://github.com/CGAL/cgal/assets/9004748/002b65fe-71e0-4237-90e6-d9254c70808d)

what I want to do next, is to filter the intersected nodes so that the number will reduce while still (almost) keep the boundary shape, like the 3rd image shown below:
![image](https://github.com/CGAL/cgal/assets/9004748/25d0263f-105c-4b85-88b0-31febe382785)

May I know if there exists a CGAL method that can handle this ? I will be grateful if anyone help this. Thank you.

**Comment by lrineau** [comment]
Hi @citystrawman, I will point out this discussion to a colleague of mine.

@janetournois Could the isotropic remeshing do the job?

**Reply by citystrawman** [reply]
> Hi @citystrawman, I will point out this discussion to a colleague of mine.
> 
> @janetournois Could the isotropic remeshing do the job?

Thanks. I tried isotropic remesh, but the result does not look good enough:
![image](https://github.com/CGAL/cgal/assets/9004748/775b3a99-e68b-4806-b4a0-b026343ef144)
![image](https://github.com/CGAL/cgal/assets/9004748/c0c88979-6087-440a-81fc-caec75000b1d)

You can see that the intersection area still keep all intersected points, which makes the shape of triangles too skiny.

This [code example](https://doc.cgal.org/latest/Polygon_mesh_processing/Polygon_mesh_processing_2manifoldness_repair_example_8cpp-example.html) inspires me, and I am trying to filter out the intersected nodes that lies on the edge of tm model, if that work, I will try to reTriangulate them(although I am not sure if it works)

**Reply by janetournois** [reply]
Hello @citystrawman ,

indeed `isotropic_remeshing()` should give better results than that.

Can you tell us which parameters you are using please?

Do you want the green boundary to be resampled but still present in the output mesh, or it does not matter?

**Reply by citystrawman** [reply]
> or it does not matter?

### `Can you tell us which parameters you are using please?`

Hi, @janetournois , I use 
```
PMP::isotropic_remeshing(faces(mesh), target_edge_length, mesh,
    CGAL::parameters::number_of_iterations(nb_iter)
    .protect_constraints(true));
``` 
in which target_edge_length=8, nb_iter = 10. 

### `Do you want the green boundary to be resampled but still present in the output mesh, or it does not matter?` 

Yes. I would want to resample the green boundary vertices (for example, remove some of the vertices that are too close) , and keep the resampled boundary lines shape (not necessarily exactly).

### `Here's my lates updates on this: `

I've been working on this problem these days and today I have some new developments. I'll describe it below:

1. Today I just found out that split() method will create duplicated intersection vertices to `tm`,  but that is not what I want(I dbout this may be the reason `isotropic_remeshing `does not work?), so I should firstly remove these duplicated vertices first;
2. after removing tm's duplicated vertices , I will use `copy_face_graph` to merge tm and splitter, however, as both splitter and tm has intersection vertices, I will have to remove these vertices again;
3. after that, I will use 
```
PMP::isotropic_remeshing(faces(mesh), target_edge_length, mesh,
    CGAL::parameters::number_of_iterations(nb_iter)
    .protect_constraints(true));
```  
to generate a remeshed model.

For previous 2 steps, everything goes fine, but the program collapse at 3rd step:
![image](https://github.com/CGAL/cgal/assets/9004748/e554ad80-bafb-497b-894c-79e498b49fe5)
![image](https://github.com/CGAL/cgal/assets/9004748/13bcf7c1-9663-4b30-8df3-2e69e8c90977)

Is it because the surface mesh has non-manifold edges that caused this collapse?

I attached my code and off file for your reference. Thank you! 

[attach.zip](https://github.com/user-attachments/files/15969005/attach.zip)

**Comment by sloriot** [comment]
Using [this example](https://doc.cgal.org/latest/Polygon_mesh_processing/Polygon_mesh_processing_2corefinement_difference_remeshed_8cpp-example.html), with 15 as target edge length in the call to  `isotropric_meshing()` (reversion the orientation of `slide.off`) I got the following:

![Screenshot from 2024-06-25 14-20-22](https://github.com/CGAL/cgal/assets/1224632/1ea2fe6c-6287-41ab-b05d-4d4759182989)

**Reply by janetournois** [reply]
`protect_constraints` must be set to false, or your constraints will not be modified at all.
If you want to keep a clean resampling (with no flips for example) of the green boundary, then you need to provide the constrained edges using the named parameter `edge_is_constrained_map` (as done in the example @sloriot pointed above)

**Reply by citystrawman** [reply]
> Using [this example](https://doc.cgal.org/latest/Polygon_mesh_processing/Polygon_mesh_processing_2corefinement_difference_remeshed_8cpp-example.html), with 15 as target edge length in the call to `isotropric_meshing()` (reversion the orientation of `slide.off`) I got the following:
> 
> ![Screenshot from 2024-06-25 14-20-22](https://private-user-images.githubusercontent.com/1224632/342722774-1ea2fe6c-6287-41ab-b05d-4d4759182989.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzMjE3MzksIm5iZiI6MTcxOTMyMTQzOSwicGF0aCI6Ii8xMjI0NjMyLzM0MjcyMjc3NC0xZWEyZmU2Yy02Mjg3LTQxYWItYjA1ZC00ZDQ3NTkxODI5ODkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjVUMTMxNzE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDhkNDgyY2E5NzYwODNmMmJiOGMwZjNiNjVmOGU0MmRjZWJiZGFlYjc4NDRiNzc2NTI4MjQ2MmVmMDE4ODNlZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.vFv3f5uGbt3YTYPoJaC7_Fn9CQrLcSH4lAlTP17qerY)

Is this one a manifold mesh or non-manifold mesh? I need a surface mesh that keep the slide inside the slope.

**Reply by citystrawman** [reply]
> `protect_constraints` must be set to false, or your constraints will not be modified at all. If you want to keep a clean resampling (with no flips for example) of the green boundary, then you need to provide the constrained edges using the named parameter `edge_is_constrained_map` (as done in the example @sloriot pointed above)

I have tried using corefinement_difference+remesh before, and corefinement_difference does not keep the upper surface of slope and only keeps the slide that is inside slope. What I want is that both the slope surface as well as the slide coexists.

**Reply by sloriot** [reply]
That is not possible, this is non-manifold and cannot be put in a Surface_mesh. You can the different part if you want but they cannot be in a closed mesh. In this context, non-manifold means that you have more than 2 triangles sharing a common edge (3 triangles in your case along the intersection polyline).

**Reply by citystrawman** [reply]
> That is not possible, this is non-manifold and cannot be put in a Surface_mesh. You can the different part if you want but they cannot be in a closed mesh. In this context, non-manifold means that you have more than 2 triangles sharing a common edge (3 triangles in your case along the intersection polyline).

is that mathematically impossible? The purple screenshot I have shown in my first post is generated by Rhino GrassHopper using exactly the same slope and slide, and I just want to simulate that by using CGAL.

**Reply by sloriot** [reply]
You can for example use `clip()` with `slope` as clipper and `slide` as mesh to be clipped. Then you can join the two meshes, but in the end you will get a closed mesh (`slide`) and one other connected component that have one boundary.

**Reply by citystrawman** [reply]
> You can for example use `clip()` with `slope` as clipper and `slide` as mesh to be clipped. Then you can join the two meshes, but in the end you will get a closed mesh (`slide`) and one other connected component that have one boundary.

If I understand it correctly, do you mean by 

`PMP::clip(slide, slope, params::default_values(), params::do_not_modify(false));` 

and then  `CGAL::copy_face_graph(slide, slope);` ?  

but if this is correct, the closed mesh should be slope?
